Move GraphML loading into a library and wrap for Python

arthurp opened the pull request:
With: https://github.com/KatanaGraph/katana-enterprise/pull/1055

arthurp commented:
You might want to call `ccache -z` to zero out the stats before you build. Otherwise your stats may be poluted by the ccache behavior during the run the generated the loaded ccache data.

BTW, I'm not totally sure ccache is working for conda, I just hadn't got around to debugging it. I wonder if we should have `ccache -z/-s` calls by default in all the builds that use ccache so it's easy to check if the ccache is working and how much gain we are getting.

ddn0 commented:
> I would recommend setting the maximum number of cached files to something on the order of the number of files compiled during the build. So maybe, for C++ builds `ccache -F 300` (set cache size in files to 300), for conda 100, and for gpu 2000. It might also be a good idea to run `ccache -c` at the end of the build. This will make sure the cache is cleaned up before it is stored away.

Seems like a reasonable change to me. It is annoying that we have to manually tweak this, but the performance anomalies when we don't seem worse than having stale ccache settings.

arthurp commented:
It looks like the Conda build is cache missing really badly. Maybe I was overly aggressive and the caches need more space to hold a few different versions of the build. Maybe conda needs 250 or 300 files instead of 100. I see you increased the other tasks similarly.

ddn0 commented:
I bumped all the max file limits a bit from your original suggestions because I was seeing very low cache hit rates. There might be some edge case about ccache object replacement policies? Anyway, now, doing a simple rebuild (from a warm cache) gets > 90% hit rate, except for conda builds. This seems reasonable enough.

I didn't look into conda builds too deeply since the don't take a long time and the main point of this change is to keep caches bounded to give predictable behavior.

arthurp committed:
Add error converter ArrowToKatana

arthurp committed:
Refactor BuildGraph to report errors. Move GraphML into libgalois.

This moves the dependency on libxml from the tools to libgalois.

arthurp committed:
Wrap GraphML load for access from Python.

arthurp committed:
Readd mongoc to open builds.

I think this was removed way back by accident when someone
(probably me) thought that mongo support was closed only.

patrickkenney9801 commented:
This looks good to me, might be good to add support for the ConvertKatana option too since that should be run after the GraphML import to convert booleans to the expected datatype and to convert time data.

arthurp commented:
> add support for the ConvertKatana option too since that should be run after the GraphML import to convert booleans to the expected datatype and to convert time data.

... Oh. I wasn't aware. What was the normal work flow for converting GraphML to Katana Form?

@bowu, do you need me to wrap `ConvertKatana` before merging? Or is this useful without it and I should merge now?
